File: wiz-shop-java/src/main/java/com/example/wizshopjava/common/helper/JwtHelper.java
package com.example.wizshopjava.common.helper;

import java.nio.charset.StandardCharsets;
import java.security.Key;
import java.util.Date;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import com.example.wizshopjava.domain.member.entity.Member;

import io.jsonwebtoken.JwtException;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.SignatureAlgorithm;
import io.jsonwebtoken.security.Keys;
import jakarta.annotation.PostConstruct;

@Component
public class JwtHelper {

	@Value("${jwt.access-token-secret-key}")
	private String accessTokenSecretKey;

	@Value("${jwt.access-token-expires-in}")
	private long accessTokenExpiresIn;

	private Key key;

	@PostConstruct
	public void init() {
		key = Keys.hmacShaKeyFor(accessTokenSecretKey.getBytes(StandardCharsets.UTF_8));
	}

	public String generateAccessToken(Member member) {
		return Jwts.builder()
			.setSubject(member.getId().toString())
			.setIssuedAt(new Date(System.currentTimeMillis()))
			.setExpiration(new Date(System.currentTimeMillis() + accessTokenExpiresIn))
			.signWith(key, SignatureAlgorithm.HS256)
			.compact();
	}

	public void validate(String accessToken) {
		try {
			Jwts.parserBuilder()
				.setSigningKey(key)
				.build()
				.parseClaimsJws(accessToken)
				.getBody()
				.getSubject();
		} catch (JwtException e) {
			throw new RuntimeException("잘못된 AccessToken 입니다.", e);
		}
	}

	public Long extractMemberId(String accessToken) {
		try {
			String subject = Jwts.parserBuilder()
				.setSigningKey(key)
				.build()
				.parseClaimsJws(accessToken)
				.getBody()
				.getSubject();
			return Long.valueOf(subject);
		} catch (Exception e) {
			throw new RuntimeException("잘못된 AccessToken 입니다.", e);
		}
	}
}

File: wiz-todo/src/main/kotlin/com/example/wiztodo/common/JwtHelper.kt
package com.example.wiztodo.common

import org.springframework.stereotype.Component

@Component
class JwtHelper {

    fun generateAccessToken(id: Long): String {
        /**
         *  1. 얘는 그냥 주어진 바를 다하면된다. 자유롭게.
         *  2. 생각의 범위가 굉장히 좁아져요. -> 관심사가 분리됐다.
         *  3. 이 안에 코드(How)가 바뀌어도 외부에 영향(변경의 전파)을 주지 않는다.
         */
        return "fake_access_token"
    }

    fun validate(accessToken: String): Boolean {
        return true
    }
}